### PR TITLE
PR for #51: Add LODES data

### PR DESCRIPTION
**Pull request by santiagohermo**
Can you review this PR for #51, @gabrieleborg? In there:

- I added [code](https://github.com/diegogentilepassaro/min_wage_rent/blob/0f975fb893fc5e5bdda5c7ce8850547add82c01a/raw/lodes/code/build.py) to download data for LODES Origin-Destination (od), Residence Area Characteristics (rac), and Workplace Area Characteristics (rac).

- I completed a run and added the files to our Google Drive ([here](https://drive.google.com/drive/u/0/folders/1DJUpaqVDLCzBampiv8z-cXUgBH2zZxWa)). You may not see all of them because they are being uploaded!

Three comments:

- As of now I downloaded the data only for 2017. If we want more years, we can simply modify [this line](https://github.com/diegogentilepassaro/min_wage_rent/blob/0f975fb893fc5e5bdda5c7ce8850547add82c01a/raw/lodes/code/build.py#L187) and run again. (fyi @diegogentilepassaro)
- The files are zipped with format `.csv.gz`. As I understand you can read this with data.table, so I didn't unzip them. If necessary, there is a function in the code to unzip them.
- I based my code on some other code online that parallelizes across cores. [This line](https://github.com/diegogentilepassaro/min_wage_rent/blob/0f975fb893fc5e5bdda5c7ce8850547add82c01a/raw/lodes/code/build.py#L184) selects the number of cores to use.
 
Let me know if you have any questions!

### CODE REVIEW

**Comment by gabrieleborg** [comment]
I will later today @santiagohermo @diegogentilepassaro 😁

File: raw/lodes/code/build.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    build_links_args = [types, states, years, work_segs, work_types]
+    urls = build_links(build_links_args)
+    log_and_print("Links created.", logger)
+
+    download_args = [[x, path, unzip] for x in urls]
+    log_and_print("\nStart downloading files.", logger)
+
+    p = multiprocessing.Pool(n_cores)
+    downloaded_files_all = p.map(download_file, download_args)
+    log_and_print("\nDownload finished.", logger)
```
Comment (gabrieleborg) [inline raw/lodes/code/build.py:61]: Very nice! Let's use these cores 💪🏽

File: raw/lodes/code/build.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+            loc = os.path.join(path, points[1], points[3], year)
+        else: 
+            loc = os.path.join(path, points[1], points[2], points[3], year)
+
+        r = requests.get(url, stream = True)
+        f = open(os.path.join(loc, fname), "wb")
+
+        for chunk in r.iter_content(chunk_size = 1024): 
+            if chunk:
+                f.write(chunk)
```
Comment (gabrieleborg) [inline raw/lodes/code/build.py:123]: Are you splitting the file in chunks to benefit from multiprocessing?

File: raw/lodes/code/build.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+    # start_end_year = [2002, 2017]
+    start_end_year = [2017, 2017]
+    data_types = ["od","rac","wac"]                                   # LODES data categories 
+    states = ["al", "ak", "az", "ar", "ca", "co", "ct", "de", "dc",   # States acronyms
+              "fl", "ga", "hi", "id", "il", "in", "ia", "ks", "ky",
+              "la", "me", "md", "ma", "mi", "mn", "ms", "mo", "mt",
+              "ne", "nv", "nh", "nj", "nm", "ny", "nc", "nd", "oh",
+              "ok", "or", "pa", "ri", "sc", "sd", "tn", "tx", "ut",
+              "vt", "va", "wa", "wv", "wi", "wy", "pr"]
```
Comment (gabrieleborg) [inline raw/lodes/code/build.py:185]: Here we can remove Puerto Rico, "pr", since we don't have any zipcode there.

**Review by gabrieleborg — APPROVED** [review]
Very neat work @santiagohermo !

Thanks for downloading the LODES data and writing the script. This is fine for me, I just asked a couple of questions for personal interest in the code but I think this is ok. I will go on and pull this request so I can start working on building indexes using LODES data.

File: raw/lodes/code/build.py (hunk 2 above)
Comment (santiagohermo) [reply raw/lodes/code/build.py:123]: This was written like this when I first copied this script from [here](https://github.com/UrbanInstitute/lodes-data-downloads/). But yeah, that's my interpretation as well!

File: raw/lodes/code/build.py (hunk 3 above)
Comment (santiagohermo) [reply raw/lodes/code/build.py]: We can also remove the associated files from the GDrive then

Comment (gabrieleborg) [reply raw/lodes/code/build.py]: I'll do it

**Pull request merged**